### DISCUSSION TITLE
Environment modules and batch managers and Buildbot (Oh my!)

### DISCUSSION THREAD

**Discussion by MatthewHambley** [buildbot/buildbot#6610] [Q&A]
I'm investigating Buildbot for testing scientific software. This raises a number of circumstances which are not obviously provided for by Buildbot. Hopefully someone knows how these things may be accomplished.

If they are already supported that is great. If they require development I would be interested in understanding what would be involved.

Firstly our cluster and supercomputer are accessed using batch managers, in the way of these things. Any command sent to the worker needs to be inserted into a batch script and submitted to the manager.

I can't see an obvious way of handling this.

Then there is the issue of environment modules which we use to, not surprisingly, set up an environment in which to compile and run our software. These are enabled by sourcing the appropriate initialisation script `. /usr/share/Modules/init/$SHELL` then loaded using a command `module load <module name>`.

There doesn't seem to be an obvious means to make this happen. I can't do it before launching the worker process as our environment defines its own python environment which will screw up the Buildbot environment.

I can't see a way to run this script as a precursor to every command rand through the worker.

I can't immediately see a neat and tidy way of putting this scripting into the `master.cfg` script.

Has anyone solved either of these problems themselves?

**Comment by tardyp** [comment]
Hello, 
You can easily achieve that by doing a custom ShellCommand.
You would just subclass the normal ShellCommand, rewrite the __init__ to override command and prefix it with whatever is needed.
Obviously this means that you can only use the text command form (and not the list form which does not go through a a shell, and thus do not need shell escapes)

**Reply by MatthewHambley** [reply]
I've had a go at creating a suitable job step but I'm running into some issues.

Because I want to retain the special capabilities of, e.g. the compile task, I don't want to just replace the straight shell step with a special shell step. What I want to do is wrap the task to be performed with module loading. Something like the decorator pattern.

I've tried just calling the wrapped tasks "run" method from within the wrapper's "run" method but that doesn't seem to work. Is there a way to get the behaviour I want?

**Reply by p12tic** [reply]
Is it possible to create the final shell command as a text file? You can send that to worker and let it run it.

**Reply by MatthewHambley** [reply]
That would probably work but I don't think it addresses the issue I'm concerned with.

As an example, and based on my understanding, the `Compile` task is derived from `ShellCommand` so it can add behaviour around understanding compiler output. Obviously I want to use that task for my compile steps. However I have things I need to do to create an execution environment and unrelated things I need to do to submit to a batch manager.

I could copy `Compile` and add all that stuff but now what happens if I want to set the environment but not submit to a batch manager? I would need a second copy modified for that. Then a third to cover the eventuality where I'm submitting but not initialising the environment.

I would much rather have a pair of wrappers which perform their tasks and then delegate to an existing class. These can be configured at run-time based on the immediate requirement. As I noted, the [decorator pattern](https://www.baeldung.com/java-decorator-pattern) is an example of how this might be achieved.

The advantage of this approach is that tasks are tasks and the modifiers needed to actually run something are modifiers needed to actually run something. The two do  not get mixed up in each other's code.

An alternative solution would be to modify the buildbot worker to support these platform specific complexities. Current the worker seems to be a straight pipe allowing commands to be run remotely. If this change were made it would be a fairly radical departure from that behaviour as it would give the worker opinions on how things should be run.

**Reply by p12tic** [reply]
I think you could create a decorator that overrides `ShellMixin.makeRemoteShellCommand` with a wrapper function that changes the behavior in a way you want. Most steps run their actions via this function.

**Comment by p12tic** [comment]
@MatthewHambley As I understand you want to have the following workflow:
 - each worker command creates a file
 - `. /usr/share/Modules/init/$SHELL` and `module load <module name>` are added to the top of the file
 - required test commands are appended to the file
 - the file is given to batch manager as input
 - the worker waits until batch manager completes running the batch file

This workflow is something what Buildbot can support, but it requires some customizations. Like @tardyp said, you would need to customize ShellCommand step, but instead of running the command outright, first send correctly formatted batch file to worker. This would require some invoking instances of `RemoteCommand` directly. See master/buildbot/steps/source/git.py for an example of a complex step that does a bunch of file operations and commands to achieve what it wants. It contains examples of all functionality that you want.
